### PR TITLE
change trend scanning fucntion and tests

### PR DESCRIPTION
**Pull request by MislavSag**
LINKED TO https://github.com/hudson-and-thames/mlfinlab/pull/280

Description

Write a function that makes trend scanning labeling function much faster.

Fixes # (issue)
Type of change

Please delete options that are not relevant.

    New feature (non-breaking change which adds functionality)

How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

    test_trend_scanning_labels.py

Test Configuration:

    Windows 10
    VSCode

Checklist:

My code follows the style guidelines of this project
I have performed a self-review of my own code
I have commented my code, particularly in hard-to-understand areas
I have made corresponding changes to the documentation
My changes generate no new warnings
I have added tests that prove my fix is effective or that my feature works
New and existing unit tests pass locally with my changes
Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by MislavSag** [comment]
How can I see why PR didn't pass the checks?

File: mlfinlab/tests/test_trend_scanning_labels.py [hunk 1]
```diff
@@ -23,9 +23,10 @@ def setUp(self):
         # Data set used for trend scanning labels
         self.eem_close = pd.read_csv(project_path + '/test_data/stock_prices.csv', index_col=0, parse_dates=[0])
         # In 2008, EEM had some clear trends
-        self.eem_close = self.eem_close['EEM'].loc[pd.Timestamp(2008, 4, 1):pd.Timestamp(2008, 10, 1)]
-
-    def test_trend_scanning_labels(self):
+        # self.eem_close = self.eem_close['EEM'].loc[pd.Timestamp(2008, 4, 1):pd.Timestamp(2008, 10, 1)]
```
Comment (proskurin) [inline mlfinlab/tests/test_trend_scanning_labels.py:26]: Can we drop this commented block of code?

File: mlfinlab/tests/test_trend_scanning_labels.py [hunk 2]
```diff
@@ -37,7 +38,7 @@ def test_trend_scanning_labels(self):
 
         # Before 2008/5/12 we had a strong positive trend
         self.assertTrue(
-            set(tr_scan_labels.loc[pd.Timestamp(2008, 1, 1):pd.Timestamp(2008, 5, 9)].bin) == set([1]))
+            set(tr_scan_labels.reindex(pd.date_range(pd.Timestamp(2008, 1, 1), pd.Timestamp(2008, 5, 9))).dropna().bin) == set([1]))
```
Comment (proskurin) [inline mlfinlab/tests/test_trend_scanning_labels.py:41]: We use reindex instead of loc because of Pandas warning?

**Comment by proskurin** [comment]
@MislavSag, thank you very much for your PR. Several things need to be done so that PR is merged into develop:
1) I've left some small comments from my side
2) Please take a look at https://travis-ci.org/github/hudson-and-thames/mlfinlab/builds/681179504. There are some code style changes which need to made in order to path pylint check
3) You need to cover with tests lines of the code which you have added (CI raises coverage error)

File: mlfinlab/tests/test_trend_scanning_labels.py (hunk 1 above)
Comment (MislavSag) [reply mlfinlab/tests/test_trend_scanning_labels.py:26]: Yes, this can be deleted.

File: mlfinlab/tests/test_trend_scanning_labels.py (hunk 2 above)
Comment (MislavSag) [reply mlfinlab/tests/test_trend_scanning_labels.py:41]: I use reindex to be compatible with pandas > 1.0.

**Comment by MislavSag** [comment]
I correct for pylints and and comments in 1), but can't figure out what's the problem with test. It returns
```
Traceback (most recent call last):
  File "C:\Users\Mislav\mlfinlab\mlfinlab\tests\test_trend_scanning_labels.py", line 34, in test_trend_scanning_labels
    tr_scan_labels = trend_scanning_labels(self.eem_close, t_events, 20)
  File "C:\ProgramData\Anaconda3\lib\site-packages\mlfinlab\labeling\trend_scanning.py", line 81, in trend_scanning_labels
    labels.loc[:, 'ret'] = price_series.loc[labels.t1].values / price_series.loc[labels.index].values - 1
  File "C:\ProgramData\Anaconda3\lib\site-packages\pandas\core\indexing.py", line 1768, in __getitem__
    return self._getitem_axis(maybe_callable, axis=axis)
  File "C:\ProgramData\Anaconda3\lib\site-packages\pandas\core\indexing.py", line 1954, in _getitem_axis
    return self._getitem_iterable(key, axis=axis)
  File "C:\ProgramData\Anaconda3\lib\site-packages\pandas\core\indexing.py", line 1595, in _getitem_iterable
    keyarr, indexer = self._get_listlike_indexer(key, axis, raise_missing=False)
  File "C:\ProgramData\Anaconda3\lib\site-packages\pandas\core\indexing.py", line 1553, in _get_listlike_indexer
    keyarr, indexer, o._get_axis_number(axis), raise_missing=raise_missing
  File "C:\ProgramData\Anaconda3\lib\site-packages\pandas\core\indexing.py", line 1655, in _validate_read_indexer
    "Passing list-likes to .loc or [] with any missing labels "
KeyError: 'Passing list-likes to .loc or [] with any missing labels is no longer supported, see https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#deprecate-loc-reindex-listlike'
```
but I don't have loc in my test code. I will wwait for checks again and come back later to see what can be done.

**Comment by proskurin** [comment]
> I correct for pylints and and comments in 1), but can't figure out what's the problem with test. It returns
> 
> ```
> Traceback (most recent call last):
>   File "C:\Users\Mislav\mlfinlab\mlfinlab\tests\test_trend_scanning_labels.py", line 34, in test_trend_scanning_labels
>     tr_scan_labels = trend_scanning_labels(self.eem_close, t_events, 20)
>   File "C:\ProgramData\Anaconda3\lib\site-packages\mlfinlab\labeling\trend_scanning.py", line 81, in trend_scanning_labels
>     labels.loc[:, 'ret'] = price_series.loc[labels.t1].values / price_series.loc[labels.index].values - 1
>   File "C:\ProgramData\Anaconda3\lib\site-packages\pandas\core\indexing.py", line 1768, in __getitem__
>     return self._getitem_axis(maybe_callable, axis=axis)
>   File "C:\ProgramData\Anaconda3\lib\site-packages\pandas\core\indexing.py", line 1954, in _getitem_axis
>     return self._getitem_iterable(key, axis=axis)
>   File "C:\ProgramData\Anaconda3\lib\site-packages\pandas\core\indexing.py", line 1595, in _getitem_iterable
>     keyarr, indexer = self._get_listlike_indexer(key, axis, raise_missing=False)
>   File "C:\ProgramData\Anaconda3\lib\site-packages\pandas\core\indexing.py", line 1553, in _get_listlike_indexer
>     keyarr, indexer, o._get_axis_number(axis), raise_missing=raise_missing
>   File "C:\ProgramData\Anaconda3\lib\site-packages\pandas\core\indexing.py", line 1655, in _validate_read_indexer
>     "Passing list-likes to .loc or [] with any missing labels "
> KeyError: 'Passing list-likes to .loc or [] with any missing labels is no longer supported, see https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#deprecate-loc-reindex-listlike'
> ```
> 
> but I don't have loc in my test code. I will wwait for checks again and come back later to see what can be done.

Probably because your Pandas is 1.0, however CI pandas requirements is not 1.0?

**Comment by Gunther-Schulz** [comment]
I took the liberty to create a patch which passes all tests and coverage locally and with Travis.

```diff
diff --git a/mlfinlab/labeling/trend_scanning.py b/mlfinlab/labeling/trend_scanning.py
index b377223..96c81c4 100644
--- a/mlfinlab/labeling/trend_scanning.py
+++ b/mlfinlab/labeling/trend_scanning.py
@@ -9,7 +9,7 @@ from numba import njit
 
 
 @njit
-def calculate_t_values(subset, min_sample_length, step):
+def calculate_t_values(subset, min_sample_length, step):    # pragma: no cover
     """
     For loop for calculating linear regression every n steps.
 
@@ -36,8 +36,8 @@ def calculate_t_values(subset, min_sample_length, step):
 
         # get coefficient and std from linear regression
         if det == 0:
-            b_mean = [np.nan]
-            b_std = [[np.nan, np.nan]]
+            b_mean = np.array([[np.nan]])
+            b_std = np.array([[np.nan, np.nan]])
         else:
             xx_inv = np.linalg.inv(xx_)
             b_mean = xx_inv @ xy_
diff --git a/requirements.txt b/requirements.txt
index a055487..a5355d9 100644
--- a/requirements.txt
+++ b/requirements.txt
@@ -1,7 +1,7 @@
 codecov==2.0.15
 coverage==4.5.2
 cvxpy==1.0.25
-numba==0.46.0
+numba==0.47.0
 numpy==1.17.3
 matplotlib==3.1.1
 pandas==0.25.3
```

**Comment by MislavSag** [comment]
With @Gunther-Schulz 's help, it passed all checks.

Can you please check, can you merge it now?

**Comment by MislavSag** [comment]
Is there something else I have to do to merge this?

**Comment by Jackal08** [comment]
Hey @MislavSag, 

We have a bit of a backlog as we have tickets and scope coming in faster than we can push code out. We are busy updating the versions to the latest and once we have figured that out, then we can start to make sure these PRs get merged.

**Comment by MislavSag** [comment]
OK, I will wait for new version and than come back to see if checks will pass.

File: mlfinlab/labeling/trend_scanning.py [hunk 3]
```diff
@@ -5,8 +5,53 @@
...
 
-from mlfinlab.structural_breaks.sadf import get_betas
+
+@njit
+def calculate_t_values(subset, min_sample_length, step):  # pragma: no cover
+    """
+    For loop for calculating linear regression every n steps.
+
+    :param subset: (np.array) subset of indecies for which we want to calculate t values
+    :return: (float) maximum t value and index of maximum t value
```
Comment (PanPip) [inline mlfinlab/labeling/trend_scanning.py:17]: Please, add a description for the `min_sample_length` variable.

Also, please start variables descriptions with a capital letter.

File: mlfinlab/labeling/trend_scanning.py [hunk 4]
```diff
@@ -5,8 +5,53 @@
...
+
+        # get coefficient and std from linear regression
+        if det == 0:
+            b_mean = np.array([[np.nan]])
+            b_std = np.array([[np.nan, np.nan]])
+        else:
+            xx_inv = np.linalg.inv(xx_)
+            b_mean = xx_inv @ xy_
+            err = y_subset - (x_subset @ b_mean)
+            b_std = np.dot(np.transpose(err), err) / (x_subset.shape[0] - x_subset.shape[1]) * xx_inv  # pylint: disable=E1136  # pylint/issues/3139
```
Comment (PanPip) [inline mlfinlab/labeling/trend_scanning.py:45]: This issue is fixed with the updated pylint, the `# pylint: disable` comment can be removed.

**Pull request closed without merging**